File: 0337-house-robber-iii/0337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int f(TreeNode* node, bool canRob,map<pair<TreeNode*,bool>,int>& mp){
        if(!node) return 0;
        if(mp.count({node,canRob})) return mp[{node,canRob}];
        int take = 0;
        if(canRob){
            take = node->val + f(node->left,false,mp) + f(node->right,false,mp);
        }
        int notTake = f(node->left,true,mp) + f(node->right,true,mp);
        return mp[{node,canRob}]=max(take,notTake);
    }
    int rob(TreeNode* root) {
    map<pair<TreeNode*,bool>,int> mp;

        if(!root) return 0;

        return f(root,true,mp);
    }
};

File: 0337-house-robber-iii/0337-house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int f(TreeNode root, HashMap<TreeNode, Integer> mp){
        if(root == null) return 0;
        if(mp.containsKey(root)) return mp.get(root);
        int take = root.val;
        if(root.left != null){
            take += f(root.left.left , mp) + f(root.left.right , mp);
        }
        if(root.right != null){
            take += f(root.right.left , mp) + f(root.right.right , mp);
        }

        int notTake = f(root.left , mp) + f(root.right , mp);
        int ans = Math.max(take , notTake);
        mp.put(root , ans);
        return ans;
    }
    public int rob(TreeNode root) {
        if(root == null) return 0;
        HashMap<TreeNode , Integer> mp = new HashMap<>();
        return f(root , mp);
    }
}